test_runner/src/main/kotlin/ftl/shard/Shard.kt

```diff
@@ -1,123 +1,60 @@
 
     // take in the XML with timing info then return list of shards
-    private fun calculateShardsByTime(testSuite: JUnitTestSuite, maxTestShards: Int) {
-
-        // read in test methods + time from XML
-        // calculate shard time
-
-        // if maxTestShards >= # of tests ... -1 sharding. 1 test per shard.
-        // tests > maxTestShards
+    fun calculateShardsByTime(runningTests: List<String>, testResult: JUnitTestResult, maxShards: Int): List<TestShard> {
 
+        val junitMap = mutableMapOf<String, Double>()
 
-        // TODO: test / rewrite this.
-        testList.sortByDescending { it.time }
-
-        val shardList = TestShards()
-        var shardCount = 0
-        var startingDuration = testList.first().time
-
-        while (testList.size > 0) {
-            if (shardCount == maxTestShards) {
-                // we have "filled" all shards, reset starting duration/position
-                startingDuration = testList.first().time
-                System.out.println("Starting duration $startingDuration")
-                shardCount = 0
+        // Create a map with information from previous junit run
+        testResult.testsuites?.forEach { testsuite ->
+            testsuite.testcases?.forEach { testcase ->
+                junitMap[testcase.name] = testcase.time.toDouble()
             }
-            if (shardList.shards.size > shardCount) {
-                shardList.shards[shardCount].testMethods.addAll(build(testList, startingDuration).testMethods)
-            } else {
-                shardList.shards.add(build(testList, startingDuration))
-            }
-            shardCount++
         }
 
-        val testShardChunks = shardList.shards
-
-        System.out.println("WITH TIME BASED SORTING")
-        var totalTime = 0.0
-        val timelist = ArrayList<Double>()
-        for (shard in shardList.shards) {
-            var shardTotal = 0.0
-            for (test in shard.testMethods) {
-                shardTotal += test.time
-            }
-            timelist.add(shardTotal)
-            totalTime += shardTotal
+        val testcases = mutableListOf<TestMethod>()
+        runningTests.forEach {
+            val time = junitMap[it] ?: 10.0
+            testcases.add(TestMethod(it, time))
         }
 
-        System.out.println("Slowest shard = " + timelist.max())
-        System.out.println("Total Time = $totalTime")
+        val testCount = testcases.size
 
-        // Ensure we don't create more VMs than requested. VM count per run should be <= testShards
-        if (testShardChunks.size > maxTestShards) {
-            fatalError("Calculated chunks $testShardChunks is > requested $maxTestShards testShards.")
-        }
-        if (testShardChunks.isEmpty()) fatalError("Failed to populate test shard chunks")
-    }
+        // If maxShards is infinite or we have more shards than tests, let's match it
+        val shardsCount = if (maxShards == -1 || maxShards > testCount) testCount else maxShards
```

Comment:
I think this is great, we should be able to drop the old chunk based sharding method.

We'll still want to add in test-methods-always-run, that could be in a different method to follow the single responsibility principle.